File: java/exercises/src/test/java/io/github/ajoz/workshop/fp/java/part1/exercise5/Exercise5Test.java
package io.github.ajoz.workshop.fp.java.part1.exercise5;

import io.github.ajoz.workshop.fp.java.tools.Function1;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;

import static io.github.ajoz.workshop.fp.java.part1.exercise5.Exercise5.applyFirst;
import static io.github.ajoz.workshop.fp.java.part1.exercise5.Exercise5.applySecond;
import static org.junit.Assert.assertEquals;

@RunWith(Enclosed.class)
public class Exercise5Test {
    public static final class Part1 {
        @Test
        public void should_apply_first_argument() {
            // given:
            final Integer first = 1;
            final Integer second = 1;
            final Integer expected = 2;

            final Function1<Integer, Function1<Integer, Integer>> curried =
                    a -> b -> a + b;

            final Function1<Integer, Integer> tested =
                    applyFirst(curried, first);

            // when:
            final Integer actual = tested.apply(second);

            // then:
            assertEquals(expected, actual);
        }
    }

    public static final class Part2 {
        @Test
        public void should_apply_second_argument() {
            // given:
            final String first = "foobar";
            final Integer second = 42;
            final String expected = "foobar42";

            final Function1<String, Function1<Integer, String>> curried =
                    string -> integer -> string + integer;

            final Function1<String, String> tested =
                    applySecond(curried, second);

            // when:
            final String actual = tested.apply(first);

            // then:
            assertEquals(expected, actual);
        }
    }

    public static final class Part4 {
        @Test
        public void should_apply_first_argument() {
            // given:
            final Integer first = 1;
            final Integer second = 1;
            final Integer expected = 2;

            final Function2<Integer, Integer, Integer> add =
                    (a, b) -> a + b;

            final Function1<Integer, Integer> tested =
                    add.applyFirst(first);

            // when:
            final Integer actual = tested.apply(second);

            // then:
            assertEquals(expected, actual);
        }
    }

    public static final class Part5 {
        @Test
        public void should_apply_second_argument() {
            // given:
            final String first = "foobar";
            final Integer second = 42;
            final String expected = "foobar42";

            final Function2<String, Integer, String> concat =
                    (string, integer) -> string + integer;

            final Function1<String, String> tested =
                    concat.applySecond(second);

            // when:
            final String actual = tested.apply(first);

            // then:
            assertEquals(expected, actual);
        }
    }
}

File: kotlin/exercises/src/test/kotlin/io/github/ajoz/workshop/fp/kotlin/part_1/exercises/exercise_5/Exercise5Test.kt
package io.github.ajoz.workshop.fp.kotlin.part_1.exercises.exercise_5

import org.junit.Test
import org.junit.experimental.runners.Enclosed
import org.junit.runner.RunWith

import org.junit.Assert.assertEquals

@RunWith(Enclosed::class)
class Exercise5Test {
    class Part1 {
        @Test
        fun `should apply only the first argument`() {
            // given:
            val first = 1
            val second = 1
            val expected = 2

            val curried: (Int) -> (Int) -> Int =
                    { a -> { b -> a + b } }

            val tested = curried.applyFirst(first)

            // when:
            val actual = tested(second)

            // then:
            assertEquals(expected, actual)
        }
    }

    class Part2 {
        @Test
        fun `should apply only the second argument`() {
            // given:
            val first = "foobar"
            val second = 42
            val expected = "foobar42"

            val curried: (String) -> (Int) -> String =
                    { string -> { integer -> string + integer } }

            val tested = curried.applySecond(second)

            // when:
            val actual = tested(first)

            // then:
            assertEquals(expected, actual)
        }
    }

    class Part4 {
        @Test
        fun `should apply only the first argument`() {
            // given:
            val first = 1
            val second = 1
            val expected = 2

            val add: (Int, Int) -> Int =
                    { a, b -> a + b }

            val tested = add.applyFirst(first)

            // when:
            val actual = tested(second)

            // then:
            assertEquals(expected, actual)
        }
    }

    class Part5 {
        @Test
        fun `should apply only the second argument`() {
            // given:
            val first = "foobar"
            val second = 42
            val expected = "foobar42"

            val concat: (String, Int) -> String =
                    { string, integer -> string + integer }

            val tested = concat.applySecond(second)

            // when:
            val actual = tested(first)

            // then:
            assertEquals(expected, actual)
        }
    }
}
